### ISSUE TITLE
no-extra-semi should prevent `if (condition);`

### ISSUE THREAD

**Issue by jacekkopecky** [eslint/eslint#13785]
**What rule do you want to change?**

no-extra-semi

**Does this change cause the rule to produce more or fewer warnings?**

more

**How will the change be implemented? (New option, new default behavior, etc.)?**

I suspect new default behaviour would be more fitting because I can't imagine a good use of `if (cond);` but I'm open to being corrected. A new option would be an alternative.

**Please provide some example code that this change will affect:**

```js
let a = 0;
if (a === 0); 
```

**What does the rule currently do for this code?**

Ignores the code, produces no warnings. It seems to be by design.

**What will the rule do after it's changed?**

Produce a warning.

**Are you willing to submit a pull request to implement this change?**

Yes but I'm very busy unfortunately so if someone does that first, I won't mind at all. I'd need guidance on whether it should be default behaviour or a new option. Default behaviour change seems easier to implement in this case.

**Comment by anikethsaha** [comment]
THanks for the issue. 

`if (a === 0)` should not be suggested as it will be parsing error and if there is any code in next line, it would come under this block (consequent). So this is an intentional behavior.

**Comment by jacekkopecky** [comment]
Please note the semicolon after the if statement. It makes it syntactically valid.

**Comment by jacekkopecky** [comment]
Looking into the implementation, I see that extra-semi is fixable but the case I'm hoping to catch here isn't (I think). Can a fixable rule have cases where it can't fix the code?

**Comment by anikethsaha** [comment]
> Please note the semicolon after the if statement. It makes it syntactically valid. It becomes a no-op.

Let me know if I am understanding the Proposal, when the code is `if(a===0) ;`,  you want to report this right? If this is so, it is intended behavior to not report this as without the semicolon will be a parsing error or wrong output. 

> Looking into the implementation, I see that extra-semi is fixable but the case I'm hoping to catch here isn't (I think). Can a fixable rule have cases where it can't fix the code?

Yes

**Comment by jacekkopecky** [comment]
Thank you @anikethsaha . The issue is with code like this:
```js
let a=0;
if (a===0);
console.log('a is 0');
```

An experienced programmer will most likely notice the lack of indentation on `console.log` and will know they typed it wrong. An inexperienced programmer will wonder why their `console.log` is firing every time, even when `a` is not 0. I've seen students do this.

**Comment by anikethsaha** [comment]
Imo, no op if statement are sometimes intentional and useful. There can be some false cases cause of this. 

Let's see other's opinions about this

**Comment by jacekkopecky** [comment]
I can see the use of no-op loops where the condition has side effects. I can't think of useful no-op if, because it can be rewritten as simply executing the condition:

```js
// instead of
if (cond);

// have
cond;
```

(I'm not sure about labels and `with` statements, but I don't care about them because I don't use them and don't teach my students to use them.)

**Comment by mdjermanovic** [comment]
It could make sense if there's `else`:

```js
if (cond); // do nothing
else doSomething();
```

**Comment by mdjermanovic** [comment]
We have this pattern in our codebase at a couple of places (though, with curly braces):

https://github.com/eslint/eslint/blob/631ae8b50e5f7975f10860e9e763b70b4f25182e/lib/rules/no-self-assign.js#L30-L40

https://github.com/eslint/eslint/blob/631ae8b50e5f7975f10860e9e763b70b4f25182e/lib/rules/operator-linebreak.js#L160-L166

https://github.com/eslint/eslint/blob/631ae8b50e5f7975f10860e9e763b70b4f25182e/lib/rules/comma-style.js#L132-L138

**Comment by jacekkopecky** [comment]
Thank you @mdjermanovic – yes, with `else` this can make sense. I'll see if I can handle it easily  in #13787 .

**Comment by jacekkopecky** [comment]
I updated #13787 to allow `if(cond);` if there is an `else` after it.

**Comment by mdjermanovic** [comment]
That would work, but I'm not sure that reporting those semicolons belongs to `no-extra-semi`.

From [the docs](https://eslint.org/docs/rules/no-extra-semi):

> Typing mistakes and misunderstandings about where semicolons are required can lead to semicolons that are unnecessary. While not technically an error, extra semicolons can cause confusion when reading code.
>
> Rule Details
>
> This rule disallows unnecessary semicolons.

The semicolon in `if(cond);` isn't _unnecessary_. The whole `if` conditional might be unnecessary, or `;` could be an error.

By the docs and the rule name, it feels its purpose is to find semis that don't affect the behavior, which isn't the case with semicolons after `if` (regardless of whether or not there's `else`).

**Comment by jacekkopecky** [comment]
Sounds right. If no-extra-semi doesn't fit, it feels like we could want something like a `no-empty-statement` rule, for where an empty statement means useless code. This could apply to `if` without `else`, `else` itself, `with`, and `for … of`, if I start from what no-extra-semi deals with.

What do you think about such a rule?

Also, is this a good place to discuss that? (I don't know where the Eslint people like to have this type of discussion)

**Comment by mdjermanovic** [comment]
I also think that `if (cond);` (without `else`) and `with (obj);` don't have a meaningful purpose.

`else;` might be a way to say that omitting the else branch wasn't an oversight.

`for (a of b);` could be a way to get the last element, or trigger some side effects in the iterator.

**Comment by jacekkopecky** [comment]
@mdjermanovic do you think I should open a new issue to suggest a new rule to cover the above? I may not be able to work on an implementation for weeks from now, if work gets in the way... 8-(

**Comment by mdjermanovic** [comment]
> Also, is this a good place to discuss that? (I don't know where the Eslint people like to have this type of discussion)

Sorry, I forgot to answer this. Yes, github issues is a perfect place to discuss new rules and other enhancements.

> do you think I should open a new issue to suggest a new rule to cover the above?

Let's wait for more opinions from other team members on whether this could be an option in an existing rule.

We could make a "possible error" rule for `if (cond);` only (`with` shouldn't be used anyway, and we already have a rule to entirely disallow `with`), or a stylistic rule that would cover all statements listed [here](https://github.com/eslint/eslint/blob/631ae8b50e5f7975f10860e9e763b70b4f25182e/lib/rules/no-extra-semi.js#L91) with options to allow some of them per user's preferences.

I think both would be useful rules, but given our [new rules policy](https://eslint.org/blog/2020/05/changes-to-rules-policies), I'm not sure if they would qualify.

Also, it's very easy to disallow `if (cond);` by configuring the [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule:

```js
/* eslint "no-restricted-syntax": ["error", {
    "selector": "IfStatement[alternate=null] > EmptyStatement.consequent",
    "message": "Unexpected empty statement."
}]*/

if (cond); // error

if (cond) doSomething(); // ok

if (cond); else doSomething(); // ok
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1zeW50YXhcIjogW1wiZXJyb3JcIiwge1xuICAgIFwic2VsZWN0b3JcIjogXCJJZlN0YXRlbWVudFthbHRlcm5hdGU9bnVsbF0gPiBFbXB0eVN0YXRlbWVudC5jb25zZXF1ZW50XCIsXG4gICAgXCJtZXNzYWdlXCI6IFwiVW5leHBlY3RlZCBlbXB0eSBzdGF0ZW1lbnQuXCJcbn1dKi9cblxuaWYgKGNvbmQpOyAvLyBlcnJvclxuXG5pZiAoY29uZCkgZG9Tb21ldGhpbmcoKTsgLy8gb2tcblxuaWYgKGNvbmQpOyBlbHNlIGRvU29tZXRoaW5nKCk7IC8vIG9rXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjUsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**Comment by mdjermanovic** [comment]
As for existing rules, my first impression was that this doesn't belong to `no-extra-semi`, but others may have a different opinion and I wouldn't be opposed to adding this option to `no-extra-semi`.

While `;` isn't a block, I think this is more related to some problems reported by the [no-empty](https://eslint.org/docs/rules/no-empty) rule.

**Comment by jacekkopecky** [comment]
Thank you, I can live with no-restricted-syntax; I wasn't familiar with it.

Also thanks for pointing out the new rules policy, I didn't spot it when I went through the contribution guidelines. I think it might be worth adding the link to the docs like [here](https://eslint.org/docs/developer-guide/contributing/rule-changes)

For what it's worth, I too think that no-empty would be easier (and more appropriate) to change to handle `if (cond); // no else`; it's currently described as empty block, but it can add empty statement and it will read just as naturally. Then this rule could be recommended.

**Comment by nzakas** [comment]
Closing, as the rule is working as intended. If anyone would like to propose a different change, please open a new issue. Thanks!
